### ISSUE TITLE
Unable to load package `charsetdetect-ae-1.1.0.1'

### ISSUE THREAD

**Issue by noughtmare** [mvdan/hint#20]
When trying to run any function in [`charsetdetect-ae`](https://hackage.haskell.org/package/charsetdetect-ae), hint gives the error:
`unhandled ELF relocation (RelA) type 42`

See [yi-editor/yi#870](https://github.com/yi-editor/yi/issues/870)

**Comment by jhance** [comment]
This seems deeply mysterious, maybe try playing around with hint by hand in like ghci and see what happens?

**Comment by mvdan** [comment]
@noughtmare it would be helpful if you could provide the smallest possible script to run (with hint from master installed) to reproduce a crash/error.

**Comment by noughtmare** [comment]
```Haskell
{-# LANGUAGE OverloadedStrings #-}
module Main where

import Data.List

import Control.Monad
import Language.Haskell.Interpreter
import Codec.Text.Detect
import Data.Maybe

main :: IO ()
main = do r <- runInterpreter testHint
          case r of
            Left err -> putStrLn $ errorString err
            Right () -> return ()
errorString :: InterpreterError -> String
errorString (WontCompile es) = intercalate "\n" (header : map unbox es)
  where
    header = "ERROR: Won't compile:"
    unbox (GhcError e) = e
errorString e = show e

testHint :: Interpreter ()
testHint = do
  set [languageExtensions := [OverloadedStrings]]
  setImports ["Codec.Text.Detect", "Prelude"]
  liftIO . print . fromJust =<< interpret "detectEncodingName \"Hello, world!\"" (as :: Maybe String)
```

**Comment by mvdan** [comment]
@noughtmare just to confirm - what GHC, hint and charsetdetect-ae are you using with that program to reproduce the crash?

**Comment by noughtmare** [comment]
`GHC 7.10.3` and `charsetdetect-ae-1.1.0.1`

**Comment by noughtmare** [comment]
The output of running `stack build && stack exec hint-charsetdetect-exe`:
```
hint-charsetdetect-exe: /home/jaro/.stack/snapshots/x86_64-linux/lts-6.9/7.10.3/lib/x86_64-linux-ghc-7.10.3/charsetdetect-ae-1.1.0.1-Eqxd4D4PX7558iLFM6s2yg/libHScharsetdetect-ae-1.1.0.1-Eqxd4D4PX7558iLFM6s2yg.a: unhandled ELF relocation(RelA) type 42

GhcException "unable to load package `charsetdetect-ae-1.1.0.1'"
```

**Comment by mvdan** [comment]
```
 $ ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.0.1
[mvdan@carbon:2] [0] [/home/mvdan]
 $ cabal --version
cabal-install version 1.24.0.0
compiled using version 1.24.0.0 of the Cabal library
 $ runghc foo.hs
"ASCII"
```

I installed `hint` from master (`cabal install`) and `charsetdetect-ae` via `cabal install charsetdetect-ae`.

This is on Arch Linux, x86_64. Cannot reproduce.

**Comment by noughtmare** [comment]
Just tested with ghc 8.0.1, and it works. So the bug has to do with ghc 7.10.3?

**Comment by noughtmare** [comment]
Also, while `runghc app/Main.hs` works `ghc app/Main.hs -o test-exe && ./test-exe` doesn't.

**Comment by gelisam** [comment]
Aha! That's a big clue. Many problems with hint have this behavior, because hint is at the boundary between interpreted code and compiled code. Instead of using hint to interpret an expression which uses charsetdetect-ae directly, try compiling a toy library which uses charsetdetect-ae, and then use hint to interpret an expression which only uses your toy library.

**Comment by noughtmare** [comment]
That gives me the same error.

**Comment by noughtmare** [comment]
I've done some further testing and `ghc app/Main.hs -o test-exe -dynamic && ./test-exe` also works.

**Comment by noughtmare** [comment]
And here is the full error message when I run `ghc app/Main.hs -o test-exe && ./test-exe`:
```
[1 of 1] Compiling Main             ( app/Main.hs, app/Main.o )
Linking test-exe ...
test-exe: /home/jaro/.cabal/lib/x86_64-linux-ghc-8.0.1/charsetdetect-ae-1.1.0.1-FvZpee1cOSl83HCwAsW2zA/libHScharsetdetect-ae-1.1.0.1-FvZpee1cOSl83HCwAsW2zA.a: unhandled ELF relocation(RelA) type 42

test-exe: Could not on-demand load symbol '_ZN17nsSBCSGroupProberC1Ev'

test-exe: /home/jaro/.cabal/lib/x86_64-linux-ghc-8.0.1/charsetdetect-ae-1.1.0.1-FvZpee1cOSl83HCwAsW2zA/libHScharsetdetect-ae-1.1.0.1-FvZpee1cOSl83HCwAsW2zA.a: unknown symbol `_ZN17nsSBCSGroupProberC1Ev'
test-exe: Could not on-demand load symbol '_ZN19nsUniversalDetectorC2Ej'

test-exe: /home/jaro/.cabal/lib/x86_64-linux-ghc-8.0.1/charsetdetect-ae-1.1.0.1-FvZpee1cOSl83HCwAsW2zA/libHScharsetdetect-ae-1.1.0.1-FvZpee1cOSl83HCwAsW2zA.a: unknown symbol `_ZN19nsUniversalDetectorC2Ej'
test-exe: Could not on-demand load symbol 'csd_close'

test-exe: /home/jaro/.cabal/lib/x86_64-linux-ghc-8.0.1/charsetdetect-ae-1.1.0.1-FvZpee1cOSl83HCwAsW2zA/libHScharsetdetect-ae-1.1.0.1-FvZpee1cOSl83HCwAsW2zA.a: unknown symbol `csd_close'
test-exe: Could not on-demand load symbol 'charsetdetectzmaezm1zi1zi0zi1zmFvZZpee1cOSl83HCwAsW2zzA_CodecziTextziDetect_detectEncodingName_closure'

GhcException "\nByteCodeLink.lookupCE\nDuring interactive linking, GHCi couldn't find the following symbol:\n  charsetdetectzmaezm1zi1zi0zi1zmFvZZpee1cOSl83HCwAsW2zzA_CodecziTextziDetect_detectEncodingName_closure\nThis may be due to you not asking GHCi to load extra object files,\narchives or DLLs needed by your current session.  Restart GHCi, specifying\nthe missing library using the -L/path/to/object/dir and -lmissinglibname\nflags, or simply by naming the relevant files on the GHCi command line.\nAlternatively, this link failure might indicate a bug in GHCi.\nIf you suspect the latter, please send a bug report to:\n  glasgow-haskell-bugs@haskell.org\n"
```

**Comment by mvdan** [comment]
@noughtmare did you end up finding anything or did you give up?

**Comment by noughtmare** [comment]
Well it's still an issue, but I don't think I can resolve it (by myself) in a reasonable amount of time. I don't have a lot of knowledge about ghc, linking or hint so it's pretty hard to debug this problem for me. I was hoping someone else would know the cause of the problem. It's not critical for my purposes (but it would be very nice to have) and using the `-dynamic` compile time flag is a workaround, but I would like to know if it is at all possible to do this without the `-dynamic` flag.

**Comment by mvdan** [comment]
I ran into similar issues recently too, where `runghc` would make my program run fine but `ghc ... && ./program` would make it segfault.

Paging @jcpetruzza who might have a pointer since he has more experience than I do.

**Comment by jcpetruzza** [comment]
The `unhandled ELF relocation(RelA) type 42` error has happened elsewhere as well. See https://github.com/DanielG/ghc-mod/issues/762 and https://github.com/gibiansky/IHaskell/issues/636. In particular https://github.com/DanielG/ghc-mod/issues/762#issuecomment-213743585 seems relevant:

> The source of this problem seems to be that gas >= 2.26 uses some new relocation types (in particular type 42, aka R_X86_64_REX_GOTPCRELX), which are not supported by ghc's linker. As a workaround, these relocation types can be disabled by invoking ghc with the flag -opta-Wa,-mrelax-relocations=no

**Comment by mvdan** [comment]
@noughtmare could you try the workaround mentioned in that thread and report back?

**Comment by noughtmare** [comment]
It works! Thank you! 

Is the issue occurring here because hint uses an outdated version of ghci? Because I can load charsetdetect-ae without a workaround with the normal ghci. Or is that not how hint works?

**Comment by noughtmare** [comment]
This is fixed by ghc 8.0.2 🎉 !

**Comment by mvdan** [comment]
Awesome, thanks for updating us.
